# Predicting Heart disease

#### The dataset on which we are working is taken from UCI machine learning repository. It is the database from Cleveland.
#### The Attribute characteristics are as follows: Categorical, Integer, Real

## Importing python libraries

In [5]:
import pandas as pd
import numpy as np
import pandas_profiling as pp

## Importing the dataset

In [17]:
xls = pd.read_excel('heartproj.xls')

## Data Description 

### First 5 rows of the Dataset

In [3]:
xls.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

### Bottom 5 rows of the Dataset

In [4]:
xls.tail(5)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

## The size of the dataset can be calculated as

In [8]:
xls.size

4242

## For Analyzing the shape of the dataset we can use

In [5]:
xls.shape

(303, 14)

#### So, there are 303 rows and 14 columns in the dataset

## The Columns in the dataset can be seen as:

In [6]:
xls.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

#### We have a total of 14 attributes available in this dataset and each attribute is important for analysis so there is no insignificant column

## Total count of different datatypes in the dataset

In [29]:
xls.dtypes.value_counts()

int64      13
float64     1
dtype: int64

## The Datatype of each Attribute we have taken is as follows:

In [9]:
xls.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

#### from the above analysis we can see that all the attributes  are of integer type except the attribute 'oldpeak' which is of float datatype

## Information of every attribute in the dataset

In [7]:
xls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


#### Important points to be observed from the above analysis are:
##### 1. A Total of 13 attributes are available
#### 2. Each attribute is having 303 non-null values
#### 3. all the attributes  are of integer type except the attribute 'oldpeak' which is of float datatype


## Removing insignificant columns

#### We have a total of 14 attributes available in this dataset and each attribute is important for analysis so there is no insignificant column

## Missing values

### Non null Values of each column in the dataset

In [10]:
xls.count()

age         303
sex         303
cp          303
trestbps    303
chol        303
fbs         303
restecg     303
thalach     303
exang       303
oldpeak     303
slope       303
ca          303
thal        303
target      303
dtype: int64

#### The number of  non null values of each attribute is 303, which is equal to the total no. of rows in the dataset.

#### Therefore, there is not even a single missing value in the dataset

## Analyzing the data having  datatype 'int'

In [13]:
xls.select_dtypes(include=['int64'])

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  slope  ca  \
0     63    1   3       145   233    1        0      150      0      0   0   
1     37    1   2       130   250    0        1      187      0      0   0   
2     41    0   1       130   204    0        0      172      0      2   0   
3     56    1   1       120   236    0        1      178      0      2   0   
4     57    0   0       120   354    0        1      163      1      2   0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...    ...  ..   
298   57    0   0       140   241    0        1      123      1      1   0   
299   45    1   3       110   264    0        1      132      0      1   0   
300   68    1   0       144   193    1        1      141      0      1   2   
301   57    1   0       130   131    0        1      115      1      1   1   
302   57    0   1       130   236    0        0      174      0      1   1   

     thal  target  
0       1       1  
1       2       1  
2       2       1  
3       2       1  
4       2       1  
..    ...     ...  
298     3       0  
299     3       0  
300     3       0  
301     3       0  
302     2       0  

[303 rows x 13 columns]

####  303 rows*13 columns have integer datatype. 

## Analyzing the data having  datatype 'float'

In [14]:
xls.select_dtypes(include=['float64'])

oldpeak
0        2.3
1        3.5
2        1.4
3        0.8
4        0.6
..       ...
298      0.2
299      1.2
300      3.4
301      1.2
302      0.0

[303 rows x 1 columns]

#### Single column has float datatype

## Report  generated using Pandas-profiling

In [4]:
pp.ProfileReport(xls)

#### According to the above report there is one duplicate rows in the dataset

## Calculating duplicate rows if any exists

In [20]:
xls.duplicated().sum()

1

In [23]:
xls.loc[xls.duplicated(keep = False), :]

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
163   38    1   2       138   175    0        1      173      0      0.0   
164   38    1   2       138   175    0        1      173      0      0.0   

     slope  ca  thal  target  
163      2   4     2       1  
164      2   4     2       1

#### Hence, row 164  is the duplicate of the row 163

## Now removing duplicate rows if it exists

In [24]:
xls.drop_duplicates(keep = 'first')

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[302 rows x 14 columns]

#### Now, we have cleaned dataset and now it has 302 rows and 14 columns

## Statistics of our data

In [26]:
xls.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

## Formatting Data

In [ ]:
## Datatypes in the following dataset are compatible with each other.